Dedup and RAII-fy the creation of a copy of CConnman::vNodes

vasild opened the pull request:
_This is a piece of https://github.com/bitcoin/bitcoin/pull/21878, chopped off to ease review._

The following pattern was duplicated in CConnman:

```cpp
lock
create a copy of vNodes, add a reference to each one 
unlock
... use the copy ... 
lock
release each node from the copy
unlock
```

Put that code in a RAII helper that reduces it to: 

```cpp
create snapshot "snap"
... use the copy ... 
// release happens when "snap" goes out of scope

vasild committed:
net: dedup and RAII-fy the creation of a copy of CConnman::vNodes

The following pattern was duplicated in CConnman:

```cpp
lock
create a copy of vNodes, add a reference to each one
unlock
... use the copy ...
lock
release each node from the copy
unlock
```

Put that code in a RAII helper that reduces it to:

```cpp
create snapshot "snap"
... use the copy ...
// release happens when "snap" goes out of scope
```

vasild committed:
net: keep reference to each node during socket wait

Create the snapshot of `CConnman::vNodes` to operate on earlier in
`CConnman::SocketHandler()`, before calling `CConnman::SocketEvents()`
and pass the `vNodes` copy from the snapshot to `SocketEvents()`.

This will keep the refcount of each node incremented during
`SocketEvents()` so that the `CNode` object is not destroyed before
`SocketEvents()` has finished.

Currently in `SocketEvents()` we only remember file descriptor numbers
(when not holding `CConnman::cs_vNodes`) which is safe, but we will
change this to remember pointers to `CNode::m_sock`.

promag commented on src/net.cpp:
```diff
@@ -1467,8 +1467,6 @@ void CConnman::SocketHandler()
         }
     }
 
-    NodesSnapshot snap{*this};
```
157674ef213e6329215d1b7eb85dacbc23c468db

This snap catches nodes added in the above `AcceptConnection` where the new snap at the beginning doesn't. Is this intentional?

vasild replied:
Hmm, good catch! No, that is not intentional.

So, the previous code did:

1. `SocketEvents()`, adding ready sockets from `vNodes` to `{recv,send,error}_set`.
2. `AcceptConnection()`, adding new entries to `vNodes`.
3. Loop over `vNodes`, for each node:
3.1. Check whether its socket is in one of `{recv,send,error}_set`. Observation - the sockets of nodes that were added in 2. will not be in any of `{recv,send,error}_set`, so for them all of `recvSet`, `sendSet` and `errorSet` will be `false`.
3.2. `InactivityCheck()` - will also be done for nodes added in 2.

The new code will skip nodes added in 2. in the 3. loop. That's the same as the old code with respect to 3.1. The only difference is that the new code will not do `InactivityCheck()` for newly accepted nodes. I think that is ok because it is unlikely that a newly accepted peer would be considered inactive. In the rare case where this might have happened in the old code and will not happen in the new code, the new code will do the check on the next call to `SocketHandler()`.

What do you think? Leave it as is or somehow ensure that new and old behave in an identical way?

promag replied:
if `InactivityCheck` for new nodes is pointless then maybe move `// Accept new connections` to the end?

vasild replied:
Added an explicit commit (a71b3f426) to show the intention. Putting the accept at the end somehow looked strange to me.

Maybe these two actions deserve to be split in separate functions (ie split `SocketHandler()`): 1. checking which of the listening sockets are ready for IO and accepting connections on them and 2. checking which of the already connected sockets are ready for IO and processing them.

promag replied:
Since you are not splitting maybe add a comment for now like `// Take a snapshot of current nodes before acceptation new ones - new nodes are irrelevant below` or something like that.

vasild replied:
Right, a comment is warranted. Added:

```cpp
    // Accept new connections. Done after taking a snapshot of the nodes because
    // `AcceptConnection()` will add new entries to `vNodes` which are unwanted
    // in the loop below because:
    // - all of `recvSet`, `sendSet` and `errorSet` will be `false` for them
    // - `InactivityCheck()` will either be a noop or will wrongly command to
    //   disconnect the node before we have tried sending or receiving anything
    //   to it.
```

jnewbery commented:
Concept ACK

vasild commented:
`644726117...92f62d9ea`: add an explicit commit that shows we intentionally don't process newly accepted nodes, as discussed in https://github.com/bitcoin/bitcoin/pull/21943#discussion_r632150326

jnewbery commented on src/net.cpp:
```diff
@@ -1451,8 +1447,10 @@ void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_s
+
     std::set<SOCKET> recv_set, send_set, error_set;
-    SocketEvents(recv_set, send_set, error_set);
+    SocketEvents(snap.Nodes(), recv_set, send_set, error_set);
```
Could we move this call to `SocketEvents()` to below the _Accept new connections_ logic? I think it's simpler to reason about if we're not holding the nodes snapshot across the _Accept new connections_ logic.

jnewbery replied:
Sorry, ignore this. I've just realized that the _Accept new connections_ logic is using the `recv_set` that's populated by `SocketEvents()`.

practicalswift commented:
Concept ACK

amitiuttarwar commented on src/net.h:
```diff
@@ -1257,6 +1276,40 @@ class CConnman
+     * RAII helper to atomically create a copy of `vNodes` and add a reference
+     * to each of the nodes. The nodes are released when this object is destroyed.
+     */
+    class NodesSnapshot
```
I'm guessing you've defined this class within `CConnman` to enable access to the private members. You could instead define the class in `net.cpp` as a standalone by making `vNodes` and `cs_vNodes` protected & declaring `friend class NodesSnapshot` here. Seems nice to keep this mechanism that's only used internally out of the header file.

diff 👇🏽
```diff 
diff --git a/src/net.cpp b/src/net.cpp
index 5fd1bf754d..6040ee86be 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -110,6 +110,39 @@ std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(cs_mapLocalHost);
 static bool vfLimited[NET_MAX] GUARDED_BY(cs_mapLocalHost) = {};
 std::string strSubVersion;

+/**
+ * RAII helper to atomically create a copy of `vNodes` and add a reference
+ * to each of the nodes. The nodes are released when this object is destroyed.
+ */
+class NodesSnapshot
+{
+    public:
+        explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}, m_nodes_copy{m_connman.vNodes}
+        {
+            LOCK(m_connman.cs_vNodes);
+            for (auto& node : m_nodes_copy) {
+                node->AddRef();
+            }
+        }
+
+        ~NodesSnapshot()
+        {
+            LOCK(m_connman.cs_vNodes);
+            for (auto& node : m_nodes_copy) {
+                node->Release();
+            }
+        }
+
+        const std::vector<CNode*>& Nodes() const
+        {
+            return m_nodes_copy;
+        }
+
+    private:
+        const CConnman& m_connman;
+        const CConnman& m_connman;
+        const std::vector<CNode*> m_nodes_copy;
+};
+
 void CConnman::AddAddrFetch(const std::string& strDest)
 {
     LOCK(m_addr_fetches_mutex);
diff --git a/src/net.h b/src/net.h
index 846c4c0ece..f39c46c166 100644
--- a/src/net.h
+++ b/src/net.h
@@ -1021,6 +1021,10 @@ public:
     /** Return true if we should disconnect the peer for failing an inactivity check. */
     bool ShouldRunInactivityChecks(const CNode& node, std::optional<int64_t> now=std::nullopt) const;

+protected:
+    std::vector<CNode*> vNodes GUARDED_BY(cs_vNodes);
+    mutable RecursiveMutex cs_vNodes;
+
 private:
     struct ListenSocket {
     public:
@@ -1153,9 +1157,7 @@ private:
     RecursiveMutex m_addr_fetches_mutex;
     std::vector<std::string> vAddedNodes GUARDED_BY(cs_vAddedNodes);
     mutable RecursiveMutex cs_vAddedNodes;
-    std::vector<CNode*> vNodes GUARDED_BY(cs_vNodes);
     std::list<CNode*> vNodesDisconnected;
-    mutable RecursiveMutex cs_vNodes;
     std::atomic<NodeId> nLastNodeId{0};
     unsigned int nPrevNodeCount{0};

@@ -1276,40 +1278,7 @@ private:
      */
     std::vector<CService> m_onion_binds;

-    /**
-     * RAII helper to atomically create a copy of `vNodes` and add a reference
-     * to each of the nodes. The nodes are released when this object is destroyed.
-     */
-    class NodesSnapshot
-    {
-    public:
-        explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}
-        {
-            LOCK(m_connman.cs_vNodes);
-            m_nodes_copy = m_connman.vNodes;
-            for (auto& node : m_nodes_copy) {
-                node->AddRef();
-            }
-        }
-
-        ~NodesSnapshot()
-        {
-            LOCK(m_connman.cs_vNodes);
-            for (auto& node : m_nodes_copy) {
-                node->Release();
-            }
-        }
-
-        const std::vector<CNode*>& Nodes() const
-        {
-            return m_nodes_copy;
-        }
-
-    private:
-        const CConnman& m_connman;
-        std::vector<CNode*> m_nodes_copy;
-    };
-
+    friend class NodesSnapshot;
     friend struct CConnmanTest;
     friend struct ConnmanTestMsg;
 };
```

jnewbery replied:
I like this. A few notes:

- the diff as shown is corrupted (contains the `+        const CConnman& m_connman;` line twice) and so doesn't apply with `git apply`.
- `vNodes` and `cs_vNodes` don't need to change to be protected. Friend classes have access to private members.
- `NodesSnapshot` can be in the unnamed namespace since it's only needed in the net.cpp translation unit.

jnewbery replied:
The `m_nodes_copy{m_connman.vNodes}` in the initializer list is accessing `vNodes` without the `cs_vNodes` lock. You could update it to take the lock:

```diff
diff --git a/src/net.cpp b/src/net.cpp
index 6040ee86be..d39aa14b20 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -117,7 +117,9 @@ std::string strSubVersion;
 class NodesSnapshot
 {
     public:
-        explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}, m_nodes_copy{m_connman.vNodes}
+        explicit NodesSnapshot(const CConnman& connman) :
+            m_connman{connman},
+            m_nodes_copy{WITH_LOCK(m_connman.cs_vNodes, return m_connman.vNodes;)}
         {
             LOCK(m_connman.cs_vNodes);
             for (auto& node : m_nodes_copy) {
```

But that would introduce a race condition if another thread deleted a node between the initializer list and the ctor being run.

I think better just to assign `m_nodes_copy` in the ctor.

vasild replied:
> I'm guessing you've defined this class within CConnman to enable access to the private members.

No, I did it to minimize the scope of the newly added class.

> Seems nice to keep this mechanism that's only used internally out of the header file.

Why? It is used internally by `CConnman` and is defined in its `private:` section. If defined in `net.cpp` that would broaden its scope unnecessary.

jnewbery replied:
> > Seems nice to keep this mechanism that's only used internally out of the header file.

> Why?

Because it's good to keep header files limited to what's required to be included into other translation units (as far as is possible). That's beneficial for both build time and code management reasons.

I think both ways are fine, and that eventually we'll want to virtualize both `CNode`'s and `CConnman`'s public interfaces so that they can be mocked out for testing, at which point the private methods could all be moved to the implementation file.

vasild replied:
> That's beneficial for both build time...

I do not think it is justified to put private class members as globals in the `.cpp` file in order to minimize the header file so that it compiles faster.

> ... and code management reasons.

Unnecessary broadened scope has an adverse effect on code management.

amitiuttarwar replied:
If `NodesSnapshot` was defined as a standalone class in `net.cpp`, it could be added to the unnamed namespace so it would only be accessible within the same translation unit, but it would broaden the scope beyond `CConnman` to within the translation unit. 

My preference for it to be in the `.cpp` is because I find it easier to understand responsibilities & maneuver the files when the headers capture the way the classes interact with external callers. But I totally understand the point about broadening scope, and think this ultimately comes down to a preference.

I'm going to resolve this conversation, thanks for the feedback @vasild & @jnewbery :)

vasild replied:
@amitiuttarwar, maybe you would like the following pattern to be more widespread:

https://github.com/bitcoin/bitcoin/blob/1ed859e90e18384376e3a1ff0cb76f3e9ab11c2d/src/txrequest.h#L96-L101

amitiuttarwar replied:
yeah, I'm def a fan of the pimpl / related patterns. Another example is how its used in `net_processing` since #20811.

amitiuttarwar commented on src/net.cpp:
```diff
@@ -1451,8 +1447,10 @@ void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_s
 
 void CConnman::SocketHandler()
 {
+    NodesSnapshot snap{*this};
```
```suggestion
    const NodesSnapshot snap{*this};
```

vasild replied:
Done.

amitiuttarwar commented on src/net.cpp:
```diff
@@ -2174,41 +2159,29 @@ void CConnman::ThreadMessageHandler()
         {
-            if (pnode->fDisconnect)
-                continue;
+            NodesSnapshot snap{*this};
```
```suggestion
            const NodesSnapshot snap{*this};
```

vasild replied:
Done.

amitiuttarwar commented on src/net.h:
```diff
@@ -1257,6 +1276,40 @@ class CConnman
+
+    private:
+        const CConnman& m_connman;
+        std::vector<CNode*> m_nodes_copy;
```
```suggestion
        const std::vector<CNode*> m_nodes_copy;
```

Can `const` this if you add the constructor assignment to the initializer list (done in diff above)

vasild replied:
The copying and `AddRef()` should be done under the lock.

promag replied:
Why `AddRef`? @amitiuttarwar suggestion can be accomplished with
```diff
         explicit NodesSnapshot(const CConnman& connman)
+            : m_nodes_copy(WITH_LOCK(connman.cs_vNodes, return connman.vNodes))
         {
-            LOCK(connman.cs_vNodes);
-            m_nodes_copy = connman.vNodes;
             for (auto& node : m_nodes_copy) {
                 node->AddRef();
             }
@@ -1305,7 +1304,7 @@ private:
         }

     private:
-        std::vector<CNode*> m_nodes_copy;
+        const std::vector<CNode*> m_nodes_copy;
     };
```

vasild replied:
Yeah :) `AddRef()` needs to be under the same lock that did the copy (without release in between) because otherwise some of the copied elements may be destroyed in the meantime and `node->` to try to dereference a dangling pointer while trying to call `AddRef()`.

It should work with something like
```cpp
explicit NodesSnapshot(const CConnman& connman)
            : m_nodes_copy(WITH_LOCK(connman.cs_vNodes, for (auto& node : connman.vNodes) { node->AddRef(); } ; return connman.vNodes))
```

but that is not worth it indeed :)

amitiuttarwar commented on src/net.h:
```diff
@@ -1257,6 +1276,40 @@ class CConnman
+        {
+            LOCK(m_connman.cs_vNodes);
+            for (auto& node : m_nodes_copy) {
+                node->Release();
```
I see that you haven't changed the behavior here, but I'm curious about the pre-existing logic: 

Is it necessary for `Release()` to be under the lock? 

It makes sense that `AddRef` has to be under the lock so there isn't a race between a thread that is trying to operate on the node, and a thread that is evaluating for disconnect. However, with `Release`, I don't think there is actually potential for a race.. if `GetRefCount` returns > 0, `DisconnectNodes` will delay processing it. And since `nRefCount` is atomic, it shouldn't be a problem if multiple threads are trying to increment / decrement at the same time. Does this reasoning seem sound? Am I missing something?

jnewbery replied:
I think this is correct. There's no need to hold `cs_vNodes` when calling `node->Release();`

Removing this lock means that the `m_connman` member is no longer required.

vasild replied:
> Is it necessary for `Release()` to be under the lock?

The lock has two effects:
* All `Release()`s happen atomically together. If removed, then other threads could observe some nodes from the snapshot released and some not yet released. I think that is ok.
* No memory reordering can happen. The `nRefCount--` on the atomic variable has the same effect because `--` is equivalent to [`fetch_sub()`](https://en.cppreference.com/w/cpp/atomic/atomic/fetch_sub) which uses [`memory_order_seq_cst`](https://en.cppreference.com/w/cpp/atomic/memory_order) by default.


Removed the lock and simplified `NodesSnapshot`, thanks!

amitiuttarwar reviewed (commented):
concept ACK, some thoughts & questions as I familiarize myself with the code

jnewbery commented on src/net.h:
```diff
@@ -1257,6 +1276,40 @@ class CConnman
+            }
+        }
+
+        const std::vector<CNode*>& Nodes() const
```
What do you think about just wrapping the m_nodes_copy `begin()` and `end()` functions:

```diff
diff --git a/src/net.cpp b/src/net.cpp
index 6040ee86be..5388af02e3 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -116,6 +116,10 @@ std::string strSubVersion;
  */
 class NodesSnapshot
 {
+    private:
+        const CConnman& m_connman;
+        const std::vector<CNode*> m_nodes_copy;
+
     public:
         explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}, m_nodes_copy{m_connman.vNodes}
         {
@@ -133,14 +137,18 @@ class NodesSnapshot
             }
         }
 
-        const std::vector<CNode*>& Nodes() const
+        using iterator = decltype(m_nodes_copy.begin());
+
+        iterator begin()
         {
-            return m_nodes_copy;
+            return m_nodes_copy.begin();
+        }
+
+        iterator end()
+        {
+            return m_nodes_copy.end();
         }
 
-    private:
-        const CConnman& m_connman;
-        const std::vector<CNode*> m_nodes_copy;
 };
```

and then the client code can use a `NodeSnapshot` in range-based loops, eg:

```diff
→ git d
diff --git a/src/net.cpp b/src/net.cpp
index 6040ee86be..5388af02e3 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -1501,7 +1509,7 @@ void CConnman::SocketHandler()
     //
     // Service each socket
     //
-    for (CNode* pnode : snap.Nodes()) {
+    for (CNode* pnode : snap) {
         if (interruptNet)
             return;
 
```

vasild replied:
I think both ways would work.

The begin/end variant seems to be a bit more code - it saves just `.Nodes()` from the callers but has two methods and is also less friendly to grepping and code navigation - now I can point my "IDE" to `Nodes()` on the line `for (CNode* pnode : snap.Nodes()) {` and see what it returns or its definition. Not so with the `for (CNode* pnode : snap) {` line.

That said, I am fine either way but have a slight preference for the `.Nodes()` variant.

jnewbery replied:
No strong preference. Just pointing out an alternative. I'll mark this as resolved.

vasild commented:
`92f62d9...00a8c94`: address review suggestions

promag reviewed (commented):
Code review ACK 00a8c943d7df3e5a58e55928a8600e062c5b4ad5.

vasild commented:
`00a8c943d7...072896b7cf`: rebase due to conflicts

jonatack commented:
Concept ACK, particularly if this can reduce `vNodes` lock contentions in these areas that are among the most frequent and long-lasting of those I'm seeing on my nodes with `-debug=lock`.

It looks like the CI stalled out on the last push. In any case, it rebases cleanly onto current master, the debug build is clean, and the unit/functional test suite is green for me locally. Reviewing.

jonatack reviewed (approved):
Tested ACK 072896b7cf5e69c11a2e6d9e7163df4e0472f504 rebased to current master. Reviewed, debug built, and ran unit tests on each commit. Ran a node with this branch rebased to master on mainnet for a day (edit: 3 days) with -debug=lock and 30-40 peers and seeing an impressive reduction in the frequent vNodes lock contentions with `grep "lock contention cs_vNodes" ~/.bitcoin/debug.log` and very few contentions with `grep "connman.cs_vNodes" ~/.bitcoin/debug.log`

Nice work!

rebroad commented:
concept ACK

vasild commented:
`072896b7cf...a4fbf1ba59`: rebase and address review suggestions

vasild commented on src/net.cpp:
```diff
@@ -1492,13 +1488,21 @@ void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_s
 
 void CConnman::SocketHandler()
 {
+    const NodesSnapshot snap{*this, false};
```
Added the comment. Some grepping showed the most common form of this is `func(/* description */ param);`

jonatack replied:
Yes. I've very recently started to switch to the `/* named_arg=*/` format for `bugprone-argument-comment` added in fa57fa1a2e7. Not sure how picky clang is about it though.

https://releases.llvm.org/12.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/bugprone-argument-comment.html.

MarcoFalke replied:
My testing revealed that the `=` is needed for clang-tidy. I am likely going to use `(/*foo_bar=*/123, ` from now on (I think clang-format "mirrors" the whitespace, so clang-format valid options are `(/*foo_bar=*/123, /*bar_foor=*/321, ` or `(/* foo_bar = */ 123, /* bar_foor = */ 321, `

jonatack replied:
Thanks, good to know.

vasild replied:
`/* bar_foor= */ 321` is also ok for clang-format (re-formatting that with clang-format leaves it unmodified). I don't care which one as long as it is consistent through the code. I will use `/*foo_bar=*/123` too.

vasild commented on src/net.cpp:
Added the comment. Some grepping showed the most common form of this is `func(/* description */ param);`

vasild commented on src/net.h:
```diff
@@ -992,8 +992,27 @@ class CConnman
-    void SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_set, std::set<SOCKET> &error_set);
+
+    /**
+     * Generate a collection of sockets that we want to be checked for readiness for IO.
```
Done.

vasild commented on src/net.cpp:
```diff
@@ -1326,57 +1326,53 @@ bool CConnman::InactivityCheck(const CNode& node) const
         recv_set.insert(hListenSocket.socket);
     }
 
-    {
```
I deliberately did the white-space change in a separate commit to ease review which would be harder if logic changes are mixed with lots of white-space changes. I extended the commit message though, so that it is easier to answer "how did we end up with those unnecessary braces".

jonatack replied:
Yes, no worries. Since some time now my `.gitconfig` has this set by default to simplify whitespace review (open to suggestions).
```
[diff]
  renames = copies
	colorMoved = dimmed-zebra
	colorMovedWs = allow-indentation-change
```

jonatack commented:
Diff-review ACK a4fbf1ba594b03379a65e780892b70bb7d4c1e4e per `git range-diff a9d0cec 072896b a4fbf1b` following my earlier full review (https://github.com/bitcoin/bitcoin/pull/21943#pullrequestreview-758400122)

promag reviewed (commented):
Code review ACK a4fbf1ba594b03379a65e780892b70bb7d4c1e4e.

vasild commented:
`a4fbf1ba59...40e0bc8de2`: rebase and switch to `/*description=*/123` following https://github.com/bitcoin/bitcoin/pull/21943#discussion_r711994001.

Sorry for the noise, I hope no more changes to this PR.

Invalidates ACKs from @jonatack, @promag.

jonatack commented:
Diff-review re-ACK 40e0bc8de29eab1aec4464e93fb42eccb9a81f5d following my earlier full review (https://github.com/bitcoin/bitcoin/pull/21943#pullrequestreview-758400122)

jonatack reviewed (approved):
re-ACK 99c1af5a8f95051e98ad4b9486eef70919ab2708 per `git range-diff 927586 40e0bc8 99c1af5` following my earlier full review https://github.com/bitcoin/bitcoin/pull/21943#pullrequestreview-758400122, only change is rebase

jnewbery reviewed (commented):
Near ACK with one suggestion for clarifying the logic sequence.

LarryRuane reviewed (commented):
Perhaps this is outside the scope of the PR, but I'd like:
```diff
     void Release()
     {
+        assert(nRefCount > 0);
         nRefCount--;
     }
```

vasild committed:
net: split CConnman::SocketHandler()

`CConnman::SocketHandler()` does 3 things:
1. Check sockets for readiness
2. Process ready listening sockets
3. Process ready connected sockets

Split the processing (2. and 3.) into separate methods to make the code
easier to grasp.

Also, move the processing of listening sockets after the processing of
connected sockets to make it obvious that there is no dependency and
also explicitly release the snapshot before dealing with listening
sockets - it is only necessary for the connected sockets part.

vasild committed:
style: remove unnecessary braces

They were needed to define the scope of `LOCK(cs_vNodes)` which was
removed in the previous commit. Re-indent in a separate commit to ease
review (use `--ignore-space-change`).

promag reviewed (commented):
Code review ACK 8589c55e2777d9fac169cfa52586d67c2a4c42b2.

laanwj commented on src/net.cpp:
```diff
@@ -1498,13 +1494,21 @@ void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_s
 
 void CConnman::SocketHandler()
 {
+    const NodesSnapshot snap{*this, /*shuffle=*/false};
```
I think with the named argument this is fine.

vasild commented on src/net.cpp:
```diff
@@ -1495,28 +1495,33 @@ void CConnman::SocketEvents(const std::vector<CNode*>& nodes, std::set<SOCKET> &
-        if (hListenSocket.socket != INVALID_SOCKET && recv_set.count(hListenSocket.socket) > 0)
-        {
-            AcceptConnection(hListenSocket);
+        if (interruptNet) {
```
You are right - the surrounding `SocketEvents()` and `SocketHandlerConnected()` both contain a similar check and would quit "quickly" if `interruptNet` becomes true.

Maybe the newly added `SocketHandlerListening()` deserves such a check too, once per each listening socket to be processed?

promag replied:
Yes, doesn't hurt I guess.

vasild replied:
Done.

promag reviewed (commented):
Code review ACK db0a01e0d1ab4462b70a44c255c602ea074b96ef, moved the early return to a loop where it makes more sense.

jonatack commented on src/net.cpp:
```diff
@@ -1497,34 +1493,29 @@ void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_s
-    if (interruptNet) return;
+    std::set<SOCKET> recv_set;
+    std::set<SOCKET> send_set;
+    std::set<SOCKET> error_set;
```
4097d14 unneeded style change, would leave as-is (and as currently exists in both of the `SocketEvents()` functions)
```diff
-    std::set<SOCKET> recv_set;
-    std::set<SOCKET> send_set;
-    std::set<SOCKET> error_set;
-
+    std::set<SOCKET> recv_set, send_set, error_set;
```

vasild replied:
I think it is good to always declare one variable per line.

LarryRuane replied:
Consider:
```cpp
struct SocketSets {
    std::set<SOCKET> recv;
    std::set<SOCKET> send;
    std::set<SOCKET> error;
}
```

vasild replied:
@LarryRuane, something like this is coming as a subsequent change from https://github.com/bitcoin/bitcoin/pull/21878 (see commit `net: introduce Sock::WaitMany()` and the structures `WaitEvents` and `WaitData` introduced in that commit).

The current combination of sets stores the sockets in "one pile of sockets that are ready for read" and "one pile of sockets that are ready for write" (notice that a given socket may be in both). It is more convenient to have just one pile of sockets each one with attached ready-for-read and ready-for-write flags.

laanwj replied:
I like the idea of abstracting this but it seems out of scope of this PR.

> It is more convenient to have just one pile of sockets each one with attached ready-for-read and ready-for-write flags.

Eventually it depends on the selection mechanism used isn't it? When it's abstracted, might as well store it in a format ready to feed to say, `poll()` or `select()` whatever is used.

vasild replied:
Yes, those 3 sets are dragged around as parameters to functions and as local variables. It is an obvious improvement to pack them together somehow. Not the purpose of this PR. Once this is PR is merged I will chop off another piece from #21878 into a smaller/manageable PR that will contain the commit which introduces such a packing `net: introduce Sock::WaitMany()`. We can discuss it there, I don't have a strong opinion on how they are packed exactly. Or if this PR is stuck we can move the discussion to #21878 where `WaitData` is introduced.

vasild replied:
> Once this is PR is merged I will chop off another piece from #21878 into a smaller/manageable PR that will contain the commit which introduces such a packing `net: introduce Sock::WaitMany()`

There are too many conflicts if I try to cherry-pick the `WaitMany()` stuff on top of bare `master` - it touches the same areas of code as the preceding commits from #21878. I chopped off those preceding commits in separate PRs instead: https://github.com/bitcoin/bitcoin/pull/23601 and https://github.com/bitcoin/bitcoin/pull/23604.

```
WaitMany | net: use Sock::WaitMany() instead of CConnman::SocketEvents()
stuff    | net: introduce Sock::WaitMany()
         | net: also wait for exceptional events in Sock::Wait()

#23601   | net: don't check if the listening socket is valid

         | scripted-diff: rename CNode::cs_hSocket to CNode::m_sock_mutex
#23604   | net: use Sock in CNode
         | fuzz: move FuzzedSock earlier in src/test/fuzz/util.h

         | net: change CreateNodeFromAcceptedSocket() to take Sock
#21879   | net: use Sock in CConnman::ListenSocket
         | net: add new method Sock::Accept() that wraps accept()
```

jonatack commented on src/net.cpp:
```diff
@@ -1497,34 +1493,29 @@ void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_s
+        // in one call. If none is ready, wait for a short while and return empty sets.
+        SocketEvents(snap.Nodes(), recv_set, send_set, error_set);
+
+        SocketHandlerConnected(snap.Nodes(), recv_set, send_set, error_set);
```
4097d14 found these comments helpful, would keep
```diff
+        // Service each socket
         SocketHandlerConnected(snap.Nodes(), recv_set, send_set, error_set);
     }
 
+    // Accept new connections
     SocketHandlerListening(recv_set);
```

vasild replied:
Done.

jonatack commented on src/net.h:
```diff
@@ -983,9 +983,51 @@ class CConnman
+                                const std::set<SOCKET>& error_set);
+
+    /**
+     * Accept an incoming connection from listening sockets that are ready for read.
```
4097d14 the existing "Accept incoming connections" comment is an easier-to-understand description for me than this one. 

Maybe:
```suggestion
     * Accept incoming connections, i.e. one from each read-ready listening socket.
```

vasild replied:
Done.

jonatack reviewed (commented):
Following up on my feedback https://github.com/bitcoin/bitcoin/pull/21943#discussion_r712022174, as of db0a01e0d1a the commits are now out of order. The last commit

    style: remove unnecessary braces

    They were needed to define the scope of `LOCK(cs_vNodes)` which was
    removed in the previous commit.

should be after the second commit `net: keep reference to each node during socket wait` (or squashed).

Almost-ACK  db0a01e0d1a with some suggestions below if you retouch.

vasild commented:
`db0a01e0d1...f52b6b2d9f`: pick nits

vasild commented on src/net.h:
```diff
@@ -1177,6 +1219,44 @@ class CConnman
+            }
+            if (shuffle) {
+                FastRandomContext rng;
+                Shuffle(m_nodes_copy.begin(), m_nodes_copy.end(), rng);
```
Done.

vasild commented on src/net.cpp:
```diff
@@ -1497,34 +1493,29 @@ void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_s
-            AcceptConnection(hListenSocket);
-        }
-    }
+        const NodesSnapshot snap{*this, /*shuffle=*/false};
```
`snap` is used as a short for `snapshot`, not in the "verb" sense. I have seen this elsewhere too. For example `zfs list -t snapshot` and `zfs list -t snap` both do the same thing.

vasild commented on src/net.cpp:
```diff
@@ -1497,34 +1493,29 @@ void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_s
-        vNodesCopy = vNodes;
-        for (CNode* pnode : vNodesCopy)
-            pnode->AddRef();
+        // Check for readiness the already connected sockets and the listening sockets
```
Done.

vasild commented on src/net.cpp:
```diff
@@ -1332,57 +1332,53 @@ bool CConnman::InactivityCheck(const CNode& node) const
 }
 
-bool CConnman::GenerateSelectSet(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_set, std::set<SOCKET> &error_set)
+bool CConnman::GenerateSelectSet(const std::vector<CNode*>& nodes, std::set<SOCKET> &recv_set, std::set<SOCKET> &send_set, std::set<SOCKET> &error_set)
```
Sometimes reviewers ask for minimal changes, without combining "real" changes with whitespace ones. So I refrained from running `clang-format` on this line. Done now.

LarryRuane reviewed (commented):
code review ACK f52b6b2d9f482353821da0ef4c485c402a396c8d
comments are optional

vasild commented on src/net.cpp:
```diff
@@ -1513,18 +1513,12 @@ void CConnman::SocketHandler()
         }
     }
 
+    const NodesSnapshot snap{*this, /*shuffle=*/false};
```
> What do you think about always shuffling?

I do not have a strong opinion. My main point with the current code is to preserve behavior - it was not shuffled before, it will not be shuffled with this PR. This PR has a different purpose.

In general, I try to avoid such unnecessary behavioral changes. _They may have very subtle effects._

Maybe it is ok to always shuffle and remove the boolean argument from the `NodesSnapshot` constructor? What do other reviewers think?

promag replied:
Slight preference to keep behavior here and simplify in a follow-up.

promag reviewed (commented):
Code review ACK f52b6b2d9f482353821da0ef4c485c402a396c8d, only format changes and comment tweaks since last review.

jonatack commented:
ACK  f52b6b2d9f482353821da0ef4c485c402a396c8d changes since last review are reordered commits and code formatting+documentation improvements
